include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,462 @@
+        oss << boost::geometry::wkt(polygon);
+        if (m_polygon_projections.find(oss.str()) != m_polygon_projections.end())
+        {
+            auto& projection = m_polygon_projections[oss.str()];
+            base = projection.base;
+            v1 = projection.v1;
+            v2 = projection.v2;
+            normal = projection.normal;
+            projected_polygon = projection.projected_polygon;
+        }
+        else
+        {
+            auto const& ring = boost::geometry::exterior_ring(polygon);
+            constexpr bool is_ring_open = boost::geometry::closure<ring_3d>::value
+                == boost::geometry::open;
+
+            if ((boost::size(ring) < 3 && is_ring_open) ||
+                (boost::size(ring) < 4 && !is_ring_open))
+            {
+                visitor.template apply<failure_few_points_on_face>();
+                return intersection_type::invalid;
+            }
+
+            // Get three non-collinear points from the polygon to define the plane
+            if (!find_non_collinear_points(polygon, base, v1, v2))
+            {
+                visitor.template apply<failure_collinear_points_on_face>();
+                return intersection_type::invalid;
+            }
+
+            v1 -= base;
+            v2 -= base;
+            normal = cross(v1, v2);
+
+            // Check for non-coplanar points
+            for (auto pit = points_begin(polygon); pit != points_end(polygon); ++pit)
+            {
+                vec3 vi;
+                set_vec_from_point::apply(*pit, vi);
+                auto vol = qvm::dot(normal, vi)- dot(normal, base);
+                if (std::abs(vol) > tolerance)
+                {
+                    visitor.template apply<failure_non_coplanar_points_on_face>();
+                    return intersection_type::invalid;
+                }
+            }
+            // Project the polygon onto the 2D plane defined by v1 and v2
+            for (auto& point : ring)
+            {
+                vec3 p;
+                set_vec_from_point::apply(point, p);
+                auto projected_p = project_point_to_plane_2d(v1, v2, p);
+                projected_polygon.outer().push_back(projected_p);
+            }
+            // Compute the orientation of the polygon and reverse it if necessary
+            if (boost::size(ring) >= 3)
+            {
+                point_2d p0 = projected_polygon.outer()[0];
+                point_2d p1 = projected_polygon.outer()[1];
+                point_2d p2 = projected_polygon.outer()[2];
+
+                auto side = strategy.side().apply(p0, p1, p2);
+
+                auto const& orientation = boost::geometry::point_order<polygon_2d>::value;
+                if ((side > 0 && orientation == boost::geometry::clockwise) ||
+                    (side < 0 && orientation == boost::geometry::counterclockwise))
+                {
+                    boost::geometry::reverse(projected_polygon);
+                }
```

Comment:
Is this implictly assuming that 
1) The polygon is convex? If not, three points may be insufficient to determine its orientation. It is not obvious to me from what I can find on the simple feature spec that it has to be.
2) Its first three points are not collinear? They may be collinear, even for a valid, convex polygon.

Reply:
Note that the correct orientation is tested by is_valid call to projected polygon below.